File: shoring/compat/src/main/java/com/panopset/compat/Splitter.java
package com.panopset.compat;

import java.util.ArrayList;
import java.util.List;

public class Splitter {
  private int i = 0;
  private final int[] w;

  private Splitter(List<Integer> widths) {
    int[] ints = new int[widths.size()];
    for (Integer x : widths) {
      ints[i++] = x;
    }
    w = ints;
  }

  private Splitter(int... widths) {
    w = widths;
  }

  public static Splitter fixedLengths(String commaSeparatedLineSplitWidths) {
    if (commaSeparatedLineSplitWidths.contains(",")) {
      List<Integer> splitWidths = new ArrayList<>();
      for (String splitWidthStr : commaSeparatedLineSplitWidths.split(",")) {
        splitWidths.add(Stringop.parseInt(splitWidthStr));
      }
      return new Splitter(splitWidths);
    }
    return new Splitter(Stringop.parseInt(commaSeparatedLineSplitWidths));
  }

  public static Splitter fixedLength(int lineSplitWidth) {
    return new Splitter(lineSplitWidth);
  }

  public static Splitter fixedLengths(int... lineSplitWidths) {
    return new Splitter(lineSplitWidths);
  }

  public List<String> split(String s) {
    List<String> rtn = new ArrayList<>();
    while (!recursivelyAccumulate(rtn, s));
    return rtn;
  }

  private boolean recursivelyAccumulate(List<String> list, String s) {
    if (s == null) {
      return true;
    }
    int nextWidth = getNextWidth();
    if (nextWidth == 0) {
      return true;
    }
    if (s.length() <= nextWidth) {
      list.add(s);
      return true;
    }
    list.add(s.substring(0, nextWidth));
    return recursivelyAccumulate(list, s.substring(nextWidth));
  }

  private int getNextWidth() {
    if (i > w.length - 1) {
      i = 0;
    }
    return w[i++];
  }
}

File: code/shoring/compat/src/main/kotlin/com/panopset/compat/Splitter.kt
package com.panopset.compat

import com.panopset.compat.Stringop.parseInt

class Splitter {
    private var i = 0
    private val w: IntArray

    constructor(widths: List<Int>) {
        val ints = IntArray(widths.size)
        for (x in widths) {
            ints[i++] = x
        }
        w = ints
    }

    constructor(vararg widths: Int) {
        w = widths
    }

    fun split(s: String?): List<String> {
        val rtn: MutableList<String> = ArrayList()
        // TODO: Wierd how it flags the more elegant empty while, and doesn't flag this clunky way of doing it.
        var done = recursivelyAccumulate(rtn, s)
        while (!done) {
            done = recursivelyAccumulate(rtn, s)
        }
        return rtn
    }

    private fun recursivelyAccumulate(list: MutableList<String>, s: String?): Boolean {
        if (s == null) {
            return true
        }
        val nextWidth = nextWidth
        if (nextWidth == 0) {
            return true
        }
        if (s.length <= nextWidth) {
            list.add(s)
            return true
        }
        list.add(s.substring(0, nextWidth))
        return recursivelyAccumulate(list, s.substring(nextWidth))
    }

    private val nextWidth: Int
        get() {
            if (i > w.size - 1) {
                i = 0
            }
            return w[i++]
        }
}

fun fixedLengths(commaSeparatedLineSplitWidths: String): Splitter {
    if (commaSeparatedLineSplitWidths.contains(",")) {
        val splitWidths: MutableList<Int> = ArrayList()
        for (splitWidthStr in commaSeparatedLineSplitWidths.split(",".toRegex()).dropLastWhile { it.isEmpty() }
            .toTypedArray()) {
            splitWidths.add(parseInt( splitWidthStr))
        }
        return Splitter(splitWidths)
    }
    return Splitter(parseInt( commaSeparatedLineSplitWidths))
}

fun fixedLength(lineSplitWidth: Int): Splitter {
    return Splitter(lineSplitWidth)
}

fun fixedLengths(vararg lineSplitWidths: Int): Splitter {
    return Splitter(*lineSplitWidths)
}
